### PR TITLE
Fix enum support

### PR DESCRIPTION
**Pull request by AlexanderSher**
- Fixes https://github.com/Azure/autorest.csharp/issues/2535
- Fixes https://github.com/Azure/autorest.csharp/issues/2556 (but tests in CADL-ranch are still required)
- Adds correct mapping between convenience and protocol method parameters
- Adds a feature to `CodeWriter` to enable pre-declaration of variables (`CodeWriter` will throw an exception if name is already available in any of the existing scopes)

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Generation/Writers/CodeWriter.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Generation/Writers/CodeWriter.cs:39]: @archerzz This is a part you can grab into the https://github.com/Azure/autorest.csharp/pull/2554 (or just wait until this one is merged).

File: src/AutoRest.CSharp/Common/Output/Models/ClientFields.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Output/Models/ClientFields.cs:105]: Not a good option, perhaps we need to make `endpoint` a known parameter,

File: src/AutoRest.CSharp/Common/Output/Models/ClientFields.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Output/Models/ClientFields.cs:120]: @archerzz This is where `ScopeDeclarations` is created to be used in scopes later

File: src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs:38]: After all the clean-ups in serialization, `Build` can be declared static.

File: src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs:16]: `RequestConditionHeaders conditionRequestFlag` isn't used in `RequestWriterHelpers`, so it is moved into `LowLevelClientMethod`

File: src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs:14]: I probably don't need this to be a record, will revert

File: test/TestProjects/CadlFirstTest/Generated/Demo2Client.cs
Comment (AlexanderSher) [inline test/TestProjects/CadlFirstTest/Generated/Demo2Client.cs:378]: @annelo-msft IMO, it is better to use a property rather than compare struct to a default value. But I can revert it, up to you.

File: src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs:146]: When method is declared, we need two scopes: inner for the body and outer that will include parameters.

File: test/TestProjects/CadlFirstTest/Generated/HelloWorldClient.cs
Comment (AlexanderSher) [inline test/TestProjects/CadlFirstTest/Generated/HelloWorldClient.cs:66]: Here are the enums, both required and optional

File: src/AutoRest.CSharp/Common/Generation/Writers/CodeWriter.cs
Comment (archerzz) [reply src/AutoRest.CSharp/Common/Generation/Writers/CodeWriter.cs:39]: Thanks. I'd like to wait for this to merge in, #2554 is not a high priority.

File: test/TestProjects/CadlFirstTest/Generated/HelloWorldClient.cs
Comment (AlexanderSher) [inline test/TestProjects/CadlFirstTest/Generated/HelloWorldClient.cs:92]: @annelo-msft We must validate all the parameters that are validated in protocol methods. Otherwise, in case of incorrect path parameter value, we will open the scope and serialize body parameter into `RequestContent` before path parameter is validated.

**Comment by annelo-msft** [comment]
> (but tests in CADL-ranch are still required)

@AlexanderSher, are you able to add these tests?

File: test/TestProjects/CadlFirstTest/Generated/Demo2Client.cs
Comment (annelo-msft) [reply test/TestProjects/CadlFirstTest/Generated/Demo2Client.cs:378]: I believe we had this discussion and decided CanBeCancelled has [too many nuances](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.canbecanceled?view=net-6.0#remarks) and what we want here is to compare it to CancellationToken.None instead of checking CanBeCancelled.  @JoshLove-msft, @tg-msft, do you remember this conversation, and can you help me confirm?

File: test/TestProjects/CadlFirstTest/Generated/HelloWorldClient.cs
Comment (annelo-msft) [reply test/TestProjects/CadlFirstTest/Generated/HelloWorldClient.cs:92]: Parameters are validated outside diagnostic scopes being opened in protocol methods, but yes, in the case we need to open the scope in the convenience method you are right.  To make it a simple rule to follow let's go ahead and validate parameters in convenience methods as well, following the same rules for parameters that are validated in protocol methods.  I'll update relevant docs.  Thanks!

File: test/TestProjects/CadlFirstTest/Generated/Demo2Client.cs
Comment (AlexanderSher) [reply test/TestProjects/CadlFirstTest/Generated/Demo2Client.cs:378]: For example, `CancellationTokenSource` uses `CanBeCancelled`: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L737

Comment (JoshLove-msft) [reply test/TestProjects/CadlFirstTest/Generated/Demo2Client.cs:378]: I thought the recommendation was to use CanBeCancelled rather than comparing to None.

File: src/AutoRest.CSharp/LowLevel/Generation/LowLevelConvenienceMethodWriter.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/LowLevel/Generation/LowLevelConvenienceMethodWriter.cs:90]: @pshao25 Variables should always be declared via `CodeWriterDeclaration`, so that `CodeWriter` can handle name conflicts

File: src/AutoRest.CSharp/LowLevel/Generation/LowLevelConvenienceMethodWriter.cs
Comment (AlexanderSher) [inline src/AutoRest.CSharp/LowLevel/Generation/LowLevelConvenienceMethodWriter.cs:105]: @pshao25 Ideally, we should compare references to parameters, i.e.: `parameter == KnownParameters.RequestContext`. If it isn't possible (e.g. in case of endpoint), we need to compare `Name` and `Type` (with or without nullability, depending on context).

**Comment by AlexanderSher** [comment]
> @AlexanderSher, are you able to add these tests?

https://github.com/Azure/cadl-ranch/pull/42

File: src/AutoRest.CSharp/LowLevel/AutoRest/DpgOutputLibrary.cs
Comment (archerzz) [inline src/AutoRest.CSharp/LowLevel/AutoRest/DpgOutputLibrary.cs:24]: So here we break the circular dependency between `ModelTypeProvider`, `TypeFactory` and `DpgOutputLibrary`?

File: src/AutoRest.CSharp/LowLevel/Generation/LowLevelClientWriter.cs
Comment (archerzz) [inline src/AutoRest.CSharp/LowLevel/Generation/LowLevelClientWriter.cs:55]: After the `scopeDeclarations`, we can invert the sequence of `WriteClientFields` and `WriteConstructors`?

File: src/AutoRest.CSharp/LowLevel/Output/DpgOutputLibraryBuilder.cs
Comment (archerzz) [inline src/AutoRest.CSharp/LowLevel/Output/DpgOutputLibraryBuilder.cs:55]: For swagger input, where do we put the model?

**Review by archerzz — APPROVED** [review]
I left some comments, but no obvious blocking issue.

File: src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs
Comment (pshao25) [inline src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs:389]: How about we call it `ParameterSet`? `ParameterSuit`?

File: src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs
Comment (AlexanderSher) [reply src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs:146]: Yes. Scopes require some rework, but it would be a big change in the code that uses it, so it will be a separate work item.

File: src/AutoRest.CSharp/LowLevel/AutoRest/DpgOutputLibrary.cs
Comment (AlexanderSher) [reply src/AutoRest.CSharp/LowLevel/AutoRest/DpgOutputLibrary.cs:24]: Kind of. I hope that when we move MPG on the CADL rails, we can fix this completely.

File: src/AutoRest.CSharp/LowLevel/Output/DpgOutputLibraryBuilder.cs
Comment (AlexanderSher) [reply src/AutoRest.CSharp/LowLevel/Output/DpgOutputLibraryBuilder.cs:55]: We don't support models for swagger input (at least for now).

File: src/AutoRest.CSharp/LowLevel/Generation/LowLevelConvenienceMethodWriter.cs
Comment (AlexanderSher) [reply src/AutoRest.CSharp/LowLevel/Generation/LowLevelConvenienceMethodWriter.cs:1]: This one has several signs of a lazy class:
- only one code path (technically two, but that is the same method called with the same parameters, just with or without scope)
- leaky encapsulation: method _assumes_ that some other code has generated protocol method that matches what is generated. Changes in protocol method generations will leak here and visa versa
- no reusability: code is used only in one place

`LowLevelClientWriter` has indeed become huge, but if we want split it by the units of work, then IMO we need to split it into "create normal methods", "create lro methods", "create pageable methods", "create LRO pageable methods" and "utilities".

File: src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs
Comment (AlexanderSher) [reply src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs:389]: `ParameterGroup` will be misleading because we have grouped parameters in HLC. `ParameterSet` doesn't represent the semantics cause the idea that these parameters not just some random parameters brought together - they have a connection between them.

**Pull request merged**
